aexpect/ops.py

```diff
@@ -0,0 +1,616 @@
+    and target can be and what behavior to expect.
+    """
+    cmd = 'cp {} {}'.format(quote(source), quote(target))
+    status, output = session.cmd_status_output(cmd)
+    if status != 0:
+        raise RuntimeError('Failed to cp {} to {}: {}'
+                           .format(source, target,
+                                   output.strip() if output else ''))
+
+
+def remove(session, path):
+    """
+    Remove a directory including its contents.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str path: directory to remove
+    :raises: :py:class:`RuntimeError` if deletion fails
+
+    Calls `rm -rf`.
+    """
+    cmd = 'rm -rf ' + quote(path)
+    status, output = session.cmd_status_output(cmd)
+    if status != 0:
+        raise RuntimeError('Failed to remove {}: {}'
+                           .format(path, output.strip() if output else ''))
+
+
+def make_tempdir(session):
+    """
+    Create a temporary directory in a safe way through a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :returns: name of temporary directory
+    :rtype: str
+    :raises: :py:class:`RuntimeError` if mktemp returns non-zero exit status
+
+    Calls `mktemp -d`, refer to `man` for more info.
+    """
+    status, output = session.cmd_status_output('mktemp -d')
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError('Failed to create temporary directory: {}'
+                           .format(output))
+    return output
+
+
+def make_tempfile(session):
+    """
+    Create a temporary file in a safe way through a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :returns: name of temporary file
+    :rtype: str
+    :raises: :py:class:`RuntimeError` if mktemp returns non-zero exit status
+
+    Calls `mktemp`, refer to `man` for more info.
+    """
+    status, output = session.cmd_status_output('mktemp')
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError('Failed to create temporary file: {}'
+                           .format(output))
+    return output
+
+
+def cat(session, filename):
+    """
+    Get contents of a text file from a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str filename: full path of file
+    :returns: file contents
+    :rtype: str
+    :raises: :py:class:`RuntimeError` if cat command fails
+
+    Should only be used for very short files without tabs or other fancy
+    contents. Otherwise better download file or use some other method.
+    """
+    cmd = 'cat ' + quote(filename)
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError('Failed to cat {}: {}'.format(filename, output))
```

Comment:
The lines between 397 and 400 are very used here in many places. One suggestion is to create a small private method/function to handle this and avoid duplication.

Reply:
The messages in the runtime errors are custom sometimes with additional information from each function. Providing a custom error message to such a refactored method is making us basically refactor just 2 lines. Also some functions perform additional status checks which fit well with the zero check. Not sure, just commenting here and I am still wondering.

Reply:
I see the "custom message" point, but still, It is not just two lines, are many lines that you are saving.

Something like this:

```python3
def _run_process(cmd, error_msg):
   status, output = session.cmd_status_output(cmd)
   if status != 0: 
      raise RuntimeError(error_msg)   
   if output:
      return output.strip()
```

Reply:
Alright, will take a note of this and comment once again after a push.